### DISCUSSION TITLE
Query related to PersistentDict

### DISCUSSION THREAD

**Discussion by animesh-workplace** [inducer/pytools#130] [Q&A]
> This module contains functionality that allows hashing with keys that remain valid across interpreter invocations, unlike Python’s built-in hashes.

Regarding the above line in the description of PersistentDict available at [link](https://documen.tician.de/pytools/persistent_dict.html). What happens when multiple instances of the same program using PersistenDict are running?
Do all the instances use:
- the same dict to update the values
- different dict to update the values

**Comment by inducer** [comment]
`PersistentDict` is safe for concurrent use, i.e. it uses proper file locking. Its intended use case are caches that are written rarely but read frequently. Writes from one process are visible to another as soon as they're completed. That said, if you're looking for something like a proper database where concurrent writes and reads are a first-class implementation goal, I would encourage you to look at redis, memcache, sqlite, or postgres.

**Reply by animesh-workplace** [reply]
Thank you for opening the discussion tab

**Reply by animesh-workplace** [reply]
My use-case for persistentDict is to use this as a database to transfer some important values across rules in snakemake. Using a proper database will be an overkill in my opinion.
